src/ore/src/id_gen.rs

```diff
@@ -126,35 +121,66 @@ where
 {
     /// Creates a new [`IdHandle`] with the provided ID.
     ///
-    /// Note: It is __entirely__ up to the caller to make sure the provided ID is not already being
-    /// used.
+    /// Note: It is *entirely* up to the caller to make sure the provided ID is
+    /// not already being used.
     pub const fn new_static(id: T) -> Self {
         IdHandle::Static(id)
     }
 
-    /// Returns the ID associated with this [`IdHandle`].
-    pub fn val(&self) -> T {
-        match self {
-            IdHandle::Static(id) => *id,
-            IdHandle::Dynamic(inner) => *inner.id(),
-        }
-    }
-
     /// Allocates a new ID and returns an owned [`IdHandle`] that can be cloned.
     fn new(allocator: &IdAllocator<T>) -> Option<Self> {
         let inner = Arc::new(internal::IdHandleInner::new(allocator)?);
         Some(IdHandle::Dynamic(inner))
     }
 }
 
+impl<T> PartialEq for IdHandle<T>
+where
+    T: PartialEq + From<u8> + AddAssign + PartialOrd + Copy,
+{
+    fn eq(&self, other: &Self) -> bool {
+        **self == **other
+    }
+}
+impl<T> Eq for IdHandle<T> where T: PartialEq + From<u8> + AddAssign + PartialOrd + Copy {}
+
+impl<T> PartialOrd for IdHandle<T>
+where
+    T: PartialOrd + From<u8> + AddAssign + Copy,
+{
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        (**self).partial_cmp(other)
+    }
+}
+
+impl<T> Ord for IdHandle<T>
+where
+    T: Ord + From<u8> + AddAssign + PartialOrd + Copy,
+{
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        (**self).cmp(other)
+    }
+}
+
 impl<T> Borrow<T> for IdHandle<T>
 where
     T: From<u8> + AddAssign + PartialOrd + Copy,
 {
     fn borrow(&self) -> &T {
+        &**self
+    }
+}
+
+impl<T> Deref for IdHandle<T>
+where
+    T: From<u8> + AddAssign + PartialOrd + Copy,
```

Comment:
I thought about implementing `Deref` for `IdHandle`, but didn't because I thought IDs should be fairly opaque, and implementing `Deref` makes it easy to use an `IdHandle` as the inner primitive type. E.g. other than serialization, we shouldn't really be concerned with the actual value of an ID, and for those cases calling `.val()` would be easy enough.

What are your thoughts @benesch? 🙂

Reply:
Makes sense, but in that case I think `val` isn't a scary enough name! I put up https://github.com/MaterializeInc/materialize/pull/19695/commits/ea8b8a93c22d81124dea1565e5fd28c54efec02c. Lemme know what ya think.